### PR TITLE
add Spotify token fetch

### PR DESCRIPTION
**Pull request by akhrail1996**
enables users to make basic API requests which do not requite user auth (search for tracks)

### CODE REVIEW

**Comment by akhrail1996** [comment]
@ruslandoga do you think that a test should be added to this PR?

File: lib/t/spotify.ex [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  @moduledoc "spotify api client"
+
+  @spotify_token_url "https://accounts.spotify.com/api/token"
+
+  def token do
+    config = Application.get_env(:t, __MODULE__)
+    token(config)
+  end
+
+  # TODO cache?
```
Comment (ruslandoga) [inline lib/t/spotify.ex:11]: cache

File: lib/t/spotify.ex [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  def token(config) do
+    client_id = Keyword.fetch!(config, :client_id)
+    client_secret = Keyword.fetch!(config, :client_secret)
+
+    auth_key = Base.encode64(client_id <> ":" <> client_secret)
+
+    headers = [
+      {"authorization", "basic " <> auth_key},
```
Comment (ruslandoga) [inline lib/t/spotify.ex:19]: Basic

File: lib/t/spotify.ex [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      {"content-type", "application/x-www-form-urlencoded"}
+    ]
+
+    body = "grant_type=client_credentials"
```
Comment (ruslandoga) [inline lib/t/spotify.ex:23]: ?

File: lib/t/spotify.ex [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      {:ok, %Finch.Response{status: _status, body: body, headers: _headers}} ->
+        case Jason.decode(body) do
+          {:ok, %{"access_token" => token}} -> {:ok, %{token: token}}
+          {:ok, %{"error" => error}} -> {:error, %{reason: error}}
+          {:ok, body} -> {:error, %{reason: body}}
+          {:error, error} -> {:error, %{reason: error}}
+        end
+
+      {:error, reason} ->
+        {:error, %{reason: reason}}
```
Comment (ruslandoga) [inline lib/t/spotify.ex:37]: ?

File: lib/t_web/channels/profile_channel.ex [hunk 5]
```diff
@@ -71,6 +71,13 @@ defmodule TWeb.ProfileChannel do
     {:reply, reply, socket}
   end
 
+  # TODO refresh after one hour
```
Comment (ruslandoga) [inline lib/t_web/channels/profile_channel.ex:74]: why?

File: lib/t_web/channels/profile_channel.ex [hunk 6]
```diff
@@ -71,6 +71,13 @@ defmodule TWeb.ProfileChannel do
...
+  def handle_in("get-spotify-token", _params, socket) do
+    token = socket.assigns[:spotify_token] || Spotify.token()
+    socket = assign(socket, spotify_token: token)
```
Comment (ruslandoga) [inline lib/t_web/channels/profile_channel.ex:77]: why is caching per user?

**Comment by ruslandoga** [comment]
@akhrail1996 what would you test?

File: lib/t/spotify.ex (hunk 4 above)
Comment (ruslandoga) [reply lib/t/spotify.ex:37]: the caller doesn't seem to handle the error in any way, also why wrap the error reason in a map?

**Comment by ruslandoga** [comment]
@akhrail1996 did you copy it from somewhere, did we have spotify before?

File: lib/t/spotify.ex (hunk 3 above)
Comment (akhrail1996) [reply lib/t/spotify.ex:23]: required

image

from https://developer.spotify.com/documentation/general/guides/authorization/client-credentials/

File: lib/t_web/channels/profile_channel.ex [hunk 7]
```diff
@@ -71,6 +71,13 @@ defmodule TWeb.ProfileChannel do
     {:reply, reply, socket}
   end
 
+  # TODO refresh after one hour
+  def handle_in("get-spotify-token", _params, socket) do
+    token = socket.assigns[:spotify_token] || Spotify.token()
```
Comment (ruslandoga) [inline lib/t_web/channels/profile_channel.ex:76]: should it be `{:ok, token}`?

File: lib/t/spotify.ex (hunk 3 above)
Comment (ruslandoga) [reply lib/t/spotify.ex:23]: but why in body?

File: lib/t_web/channels/profile_channel.ex (hunk 5 above)
Comment (akhrail1996) [reply lib/t_web/channels/profile_channel.ex:74]: token expires in 3600 based on the reply to this request

File: lib/t_web/channels/profile_channel.ex (hunk 6 above)
Comment (akhrail1996) [reply lib/t_web/channels/profile_channel.ex:77]: right, no need to

**Comment by akhrail1996** [comment]
> @akhrail1996 did you copy it from somewhere, did we have Spotify before?

no, we had Apple Music and I was partly basing on it

File: lib/t_web/channels/profile_channel.ex (hunk 5 above)
Comment (ruslandoga) [reply lib/t_web/channels/profile_channel.ex:74]: check how apns token works in apns/token.ex

**Comment by akhrail1996** [comment]
> @akhrail1996 what would you test?

Maybe just that it gets the token, but that will require adding env variables

Which afaik we cannot do because either GitHub actions won't work or we will reveal our Spotify creds. Though maybe there is a way to be able to add this test for local execution and for Github Actions to skip it?

**Comment by ruslandoga** [comment]
> Maybe just that it gets the token

gets the token from spotify or from a mock?

File: lib/t/spotify.ex (hunk 3 above)
Comment (akhrail1996) [reply lib/t/spotify.ex:23]: I am a novice at this, where should it be?

**Comment by ruslandoga** [comment]
> afaik we cannot do because either GitHub actions won't work

why wouldn't they work?

**Comment by ruslandoga** [comment]
> we will reveal our Spotify creds

do they provide sandboxing?

**Comment by ruslandoga** [comment]
> Though maybe there is a way to be able to add this test for local execution and for Github Actions to skip it?

search for tags and skipping tests in ex_unit

**Review by ruslandoga — APPROVED** [review]

**Pull request merged**